New Page Layout Picker: new design with category names on the left and thumbnails on the right

p-jackson opened the pull request:
#### Changes proposed in this Pull Request

Redesign of the new page modal that places categories on the left and layout previews on the right. See design mockups: p9Jlb4-27F-p2.

**Before**
<img width="1447" alt="Screenshot 2021-03-04 at 10 52 38 AM" src="https://user-images.githubusercontent.com/1500769/109877327-c2074000-7cd7-11eb-970e-02bb05191eb8.png">


**After**
<img width="1447" alt="Screenshot 2021-03-04 at 10 50 34 AM" src="https://user-images.githubusercontent.com/1500769/109877131-82d8ef00-7cd7-11eb-8320-ddc3db9bd656.png">


#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* `install-plugin.sh etk update/page-layout-picker-design` on sandbox
* Create a new page on your sandboxed site, the page layout picker should appear
* Test it works, you can find a checklist of things to test in #49587

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #49587

p-jackson committed:
Page template modal appears like a regular floating dialog

p-jackson committed:
Immediately select template on click and add new modal heading

p-jackson committed:
Layout shows categories on left and previews on right

p-jackson committed:
White modal background

p-jackson committed:
Increase thumbnail size and make category buttons toggles

p-jackson committed:
Modal header style

p-jackson committed:
Mobile layout uses dropdown to choose category

p-jackson committed:
Font family, sizes and colours

p-jackson committed:
Remove unused code

p-jackson committed:
Button hover and active states

p-jackson committed:
Keyboard navigable category list

p-jackson committed:
Replace most lodash usage with standard library features

p-jackson committed:
Template selector buttons have 100:65 aspect ratio

p-jackson committed:
Add descriptive layout names under layout selectors

p-jackson committed:
enable featured category of page layouts

p-jackson committed:
Full screen MacBook should have 3 layout previews per row

p-jackson committed:
Sort home pages to correct location in category list

roo2 commented:
This is looking really good, just some minor functionality and visual feedback so far:

* I'd expect the "X" button to select the blank page layout
* I'd expect the "Esc" key or clicking outside of the modal area to do the same
* In [acceptance criteria](https://github.com/Automattic/wp-calypso/issues/49743) there were two ACs which aren't met, I feel like these are optional though
> And if the user clicks on "Undo" in the topbar of the Editor [Needs tech investigation]
> Then the editor content (if any) will be removed and the Picker modal will open again allowing the user to select another layout design
* just a minor css issue, on ipad size screens there is a 1px bar at the top of the modal where page layouts are visible
![Screen Shot 2021-03-02 at 9 10 00 AM](https://user-images.githubusercontent.com/22446385/109571765-e1fffd80-7b37-11eb-86ae-a84c470d4ed2.png)
* Another minor css one, the category buttons could use a bit of padding to make them easier to click and for when the focus state is active 
![Screen Shot 2021-03-02 at 9 18 44 AM](https://user-images.githubusercontent.com/22446385/109572101-65b9ea00-7b38-11eb-8e75-703b47fcee3d.png)
* I think that using **Bold** for the selected menu items isn't ideal because it causes the text to jump a little when selected. The language picker in /me/account/ uses a coloured underline/text-decoration, I think we should use that here too 
![Screen Shot 2021-03-02 at 9 22 05 AM](https://user-images.githubusercontent.com/22446385/109572360-ccd79e80-7b38-11eb-86c2-e3b7d3c97b32.png)



Nice work on the accessibility focus states!

roo2 commented:
just a few other little design things

There is an awkward screen size between 600px and 660px where the modal does not cover the entire screen. This is probably a very rare screen size for real users, but it's not ideal.

![Screen Shot 2021-03-02 at 1 00 41 PM](https://user-images.githubusercontent.com/22446385/109590341-86de0300-7b57-11eb-9027-3f1058bb2fa0.png)

I think there should be a little bit of padding below the scrollable area, it just looks a bit cut off
![Screen Shot 2021-03-02 at 1 04 27 PM](https://user-images.githubusercontent.com/22446385/109590606-010e8780-7b58-11eb-992f-2437c8a1b736.png)

p-jackson commented:
@roo2 good points!

> - I'd expect the "X" button to select the blank page layout
> - I'd expect the "Esc" key or clicking outside of the modal area to do the same

I 100% agree. This is how it was designed to work, but it's one of the things I was going to raise in the CfT to get feedback on. I think now that the modal _looks_ more like a modal, I think the user would expect it to behave the normal way now.

> And if the user clicks on "Undo" in the topbar of the Editor [Needs tech investigation]
Then the editor content (if any) will be removed and the Picker modal will open again allowing the user to select another layout design

Personally I'm not sure about this either, trying to think of something else in GB where hitting undo which switch views. Maybe in FSE, if you hit undo would it switch you back to the previous template you were editing? Either way, I was thinking this could be implemented in another PR since it doesn't feel like it has to ship at the same time.

> - I think that using Bold for the selected menu items isn't ideal because it causes the text to jump a little when selected. The language picker in /me/account/ uses a coloured underline/text-decoration, I think we should use that here too

I'm might be able to fix the text jumping around with some line-height magic. I think you could still be right about bold being a good way to show which category is highlighted from an a11y point of view. If you don't have great eye sight the font weight difference probably won't be enough.

Screenshot without the focus right to show how selected categories are displayed:
<img width="50%" alt="Screenshot 2021-03-02 at 8 19 42 PM" src="https://user-images.githubusercontent.com/1500769/109612483-a9dad800-7b94-11eb-9371-433eb0117759.png">


@ollierozdarz What do you think about using a text decoration like underline to show which category is highlighted?


@roo2 I'll make changes for the other points you've raised though.

p-jackson committed:
Add our own focus style to menu items

autumnfjeld commented:
@roo2 @p-jackson 

> I'd expect the "X" button to select the blank page layout
I'd expect the "Esc" key or clicking outside of the modal area to do the same

This is the intended behavior per the [design and discussion](pbAok1-1QJ-p2#comment-3923).   What is the motivation for not including it in the call for testing?

autumnfjeld commented:
@eru @p-jackson 

> And if the user clicks on "Undo" in the topbar of the Editor [Needs tech investigation]
> Then the editor content (if any) will be removed and the Picker modal will open again allowing the user to select another layout design


> Personally I'm not sure about this either, trying to think of something else in GB where hitting undo which switch views. Maybe in FSE, if you hit undo would it switch you back to the previous template you were editing? Either way, I was thinking this could be implemented in another PR since it doesn't feel like it has to ship at the same time.

The [design inspiration](pbAok1-1QJ-p2#comment-3923) for this was from Genesis Blocks.   Probably right that it can be a subsequent PR, but maybe good to get design input on this @ollierozdarz.

eru commented:
@autumnfjeld It seems you are replying to me by mistake.

autumnfjeld commented:
> @autumnfjeld It seems you are replying to me by mistake.

Not a mistake, just wanted to keep you in the loop since you were part of the discussion. :)

roo2 committed:
make the close function just close the modal

ollierozdarz commented:
> What do you think about using a text decoration like underline to show which category is highlighted?

@p-jackson I wouldn't be opposed to using a simple text-decoration: underline; for the active category if it's going to increase accessibility. I think also think the `Bold` helps. 

> The design inspiration for this was from Genesis Blocks. Probably right that it can be a subsequent PR, but maybe good to get design input on this @ollierozdarz.

@autumnfjeld, I don't have any context around this decision, sorry. The link is also 404ing for me.

roo2 committed:
update hover styles and states to be similar to the language picker

roo2 committed:
prevent 1px overscroll issue. update mobile breakoint to match what is used by gutenberg. add padding to bottom of modal.

roo2 commented:
I rebased and pushed some css changes addressing my own feedback from yesterday:

* **updated hover, focus and selected styles of the page layout categories to match the styles used on the language picker modal**. I did this to avoid the issue with text jumping when the **Bold** style is applied, and to make it consistent with other styles of the site and also it gives the buttons a slightly larger clickable area.

![Screen Shot 2021-03-04 at 12 19 53 AM](https://user-images.githubusercontent.com/22446385/109819050-5e512880-7c7f-11eb-923a-17a12ce9c33b.png)

* Applied a hack to prevent the 1px strip of contents being visible above the modal when scrolling. Unfortunately it is quite a hack, I spend a good while trying to understand it but couldn't find a cleaner fix. it is related to the `.components-modal__header` using `position: sticky` and then our code changing which element is the scroll container

* Applied 24px padding to bottom of modal

* That issue with screens between 600px and 660px was caused by the page-layout modal using 660px as the cutoff for mobile whereas Gutenberg uses 600px... 🤦 I actually updated the `page-template-modal` to use 600px, this only affects the page template modal, not other calypso code. and only for screens between 600-660px. There is still an issue with horizontal scrolling on screens of around that size but I think it's acceptable because of how rare that size would be for real users. 


I also made closing the modal take you to the empty editor.

autumnfjeld commented:
> > @autumnfjeld It seems you are replying to me by mistake.
> 
> Not a mistake, just wanted to keep you in the loop since you were part of the discussion. :)

AHHH! OK, yes it was a mistake. Wrong Eru.  🤭

p-jackson committed:
Style mobile category selector like a button

p-jackson committed:
Hardcode controls to use onboarding accent colour

The user's theme was mismatching with the onboarding style; the focus
ring for a control might be blue while the hover colour might be pink.

p-jackson committed:
Adjusting category selection state styles

roo2 commented on packages/page-template-modal/src/styles/page-template-modal.scss:
```diff
@@ -1,3 +1,8 @@
+@import '~@automattic/onboarding/styles/variables';
+@import '~@automattic/typography/styles/fonts';
+@import '~@automattic/typography/styles/variables';
```
duplicate import
```suggestion
```

p-jackson replied:
One's `/typography/` and one's `/onboarding/`. The onboarding one provides the colour and the typography one I think was for the Recoleta font definition.

roo2 replied:
haha I'm blind!

roo2 commented on packages/page-template-modal/src/components/page-template-modal.js:
```diff
@@ -172,13 +113,13 @@ export default class PageTemplateModal extends Component {
 
 		// Only overwrite the page title if the template is not one of the Homepage Layouts
-		const title = isHomepageTemplate ? null : this.getTitleByTemplateSlug( name );
+		const title = template.category === 'home' ? null : template.title || '';
```
hmm templates can have multiple categories, a template can be both "_home_" and "_featured_". If a featured _home_ template is selected, this logic may not work as expected

p-jackson replied:
Ok I remember what this was about now. This was replacing a use of `_.find`. `isHomepageTemplate` used to be calculated like this:
```js
const isHomepageTemplate = _.find( this.props.templates, { name, category: 'home' } );
```
So testing `template.category === 'home'` is the correct way to replicate the existing behaviour.

However your point still stands, this doesn't work the way the comment says it will. Templates can have multiple categories, but that's stored in the **`categories`** prop, not **`category`**. None of the current templates have a `category` property, so I think this is already broken.

My guess is we don't want homepage layouts to replace the title because the title is probably already set correctly to the site's actual title. Seems reasonable, so I'll fix/add-this-behaviour. I'll search through all the categories looking for `'home'`.

p-jackson replied:
I've fixed the "don't set the title for home layouts" behaviour. Added tests for the behaviour since it seems like the sort of thing that'd be overlooked when future changes are made.

roo2 commented on packages/page-template-modal/src/components/page-template-modal.js:
```diff
@@ -314,40 +192,20 @@ export default class PageTemplateModal extends Component {
-			? unfilteredGroups
-			: omit( unfilteredGroups, 'home-page' );
+	getTemplateCategories = () => {
+		const groups = this.getTemplateGroups();
```
nice, this function is so much clearer <3

roo2 commented on packages/page-template-modal/src/components/page-template-modal.js:
```diff
@@ -391,107 +245,112 @@ export default class PageTemplateModal extends Component {
-				isDismissible={ false }
+				onRequestClose={ this.closeModal }
+				aria={ {
+					// TODO: `labelledby` option isn't working because of a <Modal> bug in Gutenberg: WordPress/gutenboarding#29020
```
it looks like you've got this merged so you can remove the comment now (congrats) 
```suggestion
```

p-jackson replied:
Fair. It won't be fixed until GB 10.1, but there's nothing really left "todo"

roo2 commented on packages/page-template-modal/src/components/page-template-modal.js:
```diff
@@ -391,107 +245,112 @@ export default class PageTemplateModal extends Component {
+				onRequestClose={ this.closeModal }
+				aria={ {
+					// TODO: `labelledby` option isn't working because of a <Modal> bug in Gutenberg: WordPress/gutenboarding#29020
+					labelledby: `page-template-modal__heading-${ instanceId }`,
```
It looks like the `instanceId` is to ensure that we don't have duplicate IDs added to the page, since there should only be one instance of the modal, is this neccessary?

p-jackson replied:
This seems to be the idiom used in Gutenberg when you need to have a unique id for aria attributes. You're right that we could get away without it because we don't plan to have multiple copies of this component at the same time. On the other hand it seems like good practice to always ensure IDs are unique, especially when the component is in a package. If it's in a package it feels like the code should be as generic as possible?

roo2 replied:
> If it's in a package it feels like the code should be as generic as possible

Yea that makes sense to follow the conventions 👍

roo2 commented on packages/page-template-modal/src/utils/test/group-utils.js:
```diff
@@ -0,0 +1,41 @@
+/**
```
thank you for adding this test!

p-jackson replied:
np :) I wanted to remove most of our lodash uses, but wasn't confident to do the refactor without writing the tests first :D

roo2 commented on packages/page-template-modal/src/styles/page-template-modal.scss:
```diff
@@ -37,92 +31,61 @@ $wp-org-sidebar-collapsed: 36px;
 $wp-org-admin-bar-full: 32px;
```
these `$wp-org-` variables all seem to be unused

roo2 commented on packages/page-template-modal/src/styles/page-template-modal.scss:
```diff
@@ -217,234 +163,140 @@ body.admin-bar:not( .is-fullscreen-mode ) .page-template-modal-screen-overlay {
-	.block-iframe-preview__template-title {
-		padding-top: 20px;
-	}
+.page-template-modal__loading {
```
this seems to be unused now

roo2 commented on packages/page-template-modal/src/styles/page-template-modal.scss:
```diff
@@ -217,234 +163,140 @@ body.admin-bar:not( .is-fullscreen-mode ) .page-template-modal-screen-overlay {
-	padding-right: 24px;
+		position: fixed;
+		width: calc( #{$sidebar-width} - 24px );
+		pointer-events: none;
```
i don't understand the need to remove pointer events and then add it back on the children, is this still neccessary?

p-jackson replied:
Ah yeah, I should have had a comment here. This is why I did it, but let me know if you think if the behaviour isn't worth the added complexity.

I want the scrollbar for the layout list to behave as if it's the scrollbar for the whole modal (rather than a child scrollbar that's just for one of the columns inside the modal). This is mostly achieved by giving the category sidebar a `position: fixed`.
However if you use the mouse wheel while hovered over the sidebar it won't scroll the modal's scrollbar, which is the UX I was looking for. So I'm setting all the pointer events manually since the default browser behaviour wasn't giving me what I wanted. Disabling pointer events on the sidebar will allow the scroll wheel to work while moused over the sidebar, and then I need to re-enable pointer events again so that the clicks actually go to the buttons.

I'd recommend commenting out the `pointer-events` locally to get the idea of what I'm talking about.

roo2 replied:
ahhh ok, TIL about this scrolling behaviour edge case and workaround 👍 yea it would probably be worth adding a comment to the code about that. Nice work being able to address it!

roo2 reviewed (commented):
Excellent work on simplifying the code and accessibility, I have some mostly minor feedback.

p-jackson committed:
Remove TODO comment now it's complete

p-jackson committed:
Delete unused CSS

p-jackson committed:
Use blacks/grays for button styles, use dashboard theme for focus rings

p-jackson committed:
Potential fix for link text colour on Atomic

p-jackson committed:
Comment explaining when pointer-events is being manually configured

p-jackson committed:
Fix logic for not setting page title for Home layouts

roo2 reviewed (approved):
retested that *featured* *home* categories do not have their title set 👍 

Nice work on adding the enzyme test! <3 

The selected and target states look good and use the admin-theme colours 👍 

And thanks for adding that comment explaining pointer-events.

I wasn't fully across the issue with highlight colour on atomic sites, but I tested the latest code on an atomic site and with an admin theme and all was coloured as expected 👍 

LGTM 🚀

p-jackson committed:
Tidy focus ring on chromium browsers

p-jackson committed:
Remove unnecessary <fieldset> element
